drivers/linstorvolumemanager.py

```diff
@@ -196,17 +202,88 @@ def _get_controller_uri():
         # Not found, maybe we are trying to create the SR...
         pass
 
-def get_controller_uri():
-    retries = 0
-    while True:
-        uri = _get_controller_uri()
+
+@contextlib.contextmanager
+def shared_reader(path):
+    with open(path, 'r') as f:
+        lock = flock.ReadLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+@contextlib.contextmanager
+def excl_writer(path):
+    with open(path, 'r+') as f:
+        lock = flock.WriteLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+def get_cached_controller_uri(ctx=None):
+    try:
+        with ctx if ctx else shared_reader(CONTROLLER_CACHE_PATH) as f:
+            return f.read().strip()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to read controller URI cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH,e))
+
+
+def delete_controller_uri_cache(ctx=None):
+    try:
+        with ctx if ctx else excl_writer(CONTROLLER_CACHE_PATH) as f:
+            f.seek(0)
+            f.truncate()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to delete uri cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
+
+
+def write_controller_uri_cache(uri, ctx=None):
+    try:
+        if not os.path.exists(CONTROLLER_CACHE_DIRECTORY):
+            os.makedirs(CONTROLLER_CACHE_DIRECTORY)
+            os.chmod(CONTROLLER_CACHE_DIRECTORY, 0o700)
+        with ctx if ctx else excl_writer(CONTROLLER_CACHE_PATH) as f:
+            f.seek(0)
+            f.write(uri)
+            f.truncate()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to write URI cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
+
+
+def build_controller_uri_cache():
+    with excl_writer(CONTROLLER_CACHE_PATH) as f:
+        uri = get_cached_controller_uri(contextlib.nullcontext(f))
         if uri:
             return uri
+        uri = _get_controller_uri()
+        if not uri:
+            for retries in range(9):
+                time.sleep(1)
+                uri = _get_controller_uri()
+                if uri:
+                    break
 
-        retries += 1
-        if retries >= 10:
-            break
-        time.sleep(1)
+        if uri:
+            write_controller_uri_cache(uri, contextlib.nullcontext(f))
+        return uri
+
+
+def get_controller_uri():
+    uri = get_cached_controller_uri()
+    if not uri:
+        uri = build_controller_uri_cache()
+    return uri
```

Comment:
```suggestion
def get_controller_uri():
    uri = read_controller_uri_cache()
    if not uri:
        uri = build_controller_uri_cache()
    return uri
```

Reply:
done in [4f84f23](https://github.com/xcp-ng/sm/pull/83/commits/4f84f2365e9c7ef63fedc9fbdf845b498c73a2ba)